hips/hip-9999.md

```diff
@@ -0,0 +1,76 @@
+---
+hip: 9999
+title: "Document backwards-compatibility rules"
+authors: [ "Marc Khouzam <marc.khouzam@montreal.ca>" ]
+created: "2020-09-18"
+type: "process"
+status: "draft"
+---
+
+## Abstract
+
+Define and document the backwards-compatibility rules that apply to Helm minor and patch releases.
+
+## Motivation
+
+Helm aims to strictly respect [Semantic Versioning][semver], where minor and patch releases must be 100% backwards-compatible.
+Backwards-compatibility implies that when moving to a new backwards-compatible release:
+* tools (such as Continuous Integration tools) using the helm CLI will not require any changes
+* programs using the Helm Go API will continue to compile 
+
+Some contributors to the Helm code base may not be comfortable with the details of what constitutes a compatibility-breaking change.
+Furthermore, in some cases, there can be a difference of opinion, even amongst the maintainers, on what constitutes such a change.
+
+Having a well-defined, documented set of backwards-compatibility rules for the Helm code-base will allow to:
+* avoid compatibility-breaking changes being merged by mistake
+* guide contributors in understanding helm's backwards-compatibility requirements
+* remove any uncertainty when it comes to enforcing backwards compatibility
+* provide a basis for discussion and possible updates to how helm handles backwards compatibility
+
+## Rationale
+
+The maintainers have agreed that a formal definition of Helm's backwards-compatibility rules would be beneficial.
+
+## Specification
+
+The following compatibility rules will apply to helm minor and patch releases:
+* exported Go APIs will remain compatible as per this [article][go-module-comp]
+* CLI commands:
+    * are case-sensitive
+    * must not be removed
+    * must not be renamed
+    * must not be moved elsewhere in the command hierarchy
+* CLI flags:
+    * are case-sensitive
+    * must not be removed
+    * must not be renamed
+    * must not be moved in the command hierarchy
+    * must not change type
+* CLI output must:
```

Comment:
The rules to this point:

- Help text may change
- Format of structured output (tables, JSON, YAML) must not change
- Errors can change if doing so increases their usability
- Warnings and errors are sent to STDERR
- User-facing data is sent to STDOUT
- Any failure must generate the return code `1`, and any success must generate the code `0` on systems that support return codes. Warnings MUST NOT generate a `1`.
- There has never been a consensus on messages that are user-facing and informal or non-informational (e.g. "Happy Helming"). I most often vote against these changes, while others have voted for them.
- When a message contains a "bug" (including spelling errors, misinformation, or egregious grammar errors), it can be fixed
- In the past, we have allowed formatting changes to structured formats (JSON, YAML) when such changes were not considered _structural_ by the relevant specification. E.g. adding/removing whitespace in JSON documents. These days, we have been more resistant to that... but we have never explicitly said it is not allowed
- We do need to say something about "log" messages. We currently have no policy at all about log messages, and hence allow any changes to messages at the Debug and Warning levels.

Reply:
> The rules to this point:
> * Help text may change
> * Format of structured output (tables, JSON, YAML) must not change
> * Errors can change if doing so increases their usability
> * When a message contains a "bug" (including spelling errors, misinformation, or egregious grammar errors), it can be fixed

Added

> * Warnings and errors are sent to STDERR
> * User-facing data is sent to STDOUT
> * Any failure must generate the return code 1, and any success must generate the code 0 on systems that support return codes. Warnings MUST NOT generate a 1.

These are design rules.
Although I added that that return codes should not be changed unless they were incorrect.

> * There has never been a consensus on messages that are user-facing and informal or non-informational (e.g. "Happy Helming"). I most often vote against these changes, while others have voted for them.
> * In the past, we have allowed formatting changes to structured formats (JSON, YAML) when such changes were not considered structural by the relevant specification. E.g. adding/removing whitespace in JSON documents. These days, we have been more resistant to that... but we have never explicitly said it is not allowed
> * We do need to say something about "log" messages. We currently have no policy at all about log messages, and hence allow any changes to messages at the Debug and Warning levels.

I've added this to the open issues section for more discussion